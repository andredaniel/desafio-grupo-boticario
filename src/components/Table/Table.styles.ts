import styled from 'styled-components'
import { theme } from '../../theme'

export const TableWrapper = styled.div`
  table {
    border-collapse: collapse;
    white-space: nowrap;
    width: 100%;

    thead {
      color: ${theme.colors.secondary.LIGHT};
      text-align: left;

      th {
        font-size: 0.85rem;
      }
    }

    td,
    th {
      padding: 15px 20px;
    }

    tbody {
      tr {
        padding: 0;

        &:nth-child(odd) {
          background-color: #f2f5faaa;

          td:first-child {
            border-bottom-left-radius: 10px;
            border-top-left-radius: 10px;
          }

          td:last-child {
            border-bottom-right-radius: 10px;
            border-top-right-radius: 10px;
          }
        }
      }
    }
  }
`
