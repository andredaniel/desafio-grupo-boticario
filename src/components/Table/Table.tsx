import React from 'react'
import Skeleton from 'react-loading-skeleton'
import { TableWrapper } from './Table.styles'

interface Props {
  loading?: boolean
}

const createArray = (length: number) => {
  const arr = []

  for (let index = 0; index < length; index++) {
    arr.push(index)
  }

  return arr
}

const Table: React.FC<Props> = ({ children, loading }): JSX.Element => {
  const child: any = React.Children.toArray(children)[0]

  const rowTimes: number[] = createArray(
    Math.ceil((window.innerHeight - 160) / 80)
  )

  const colTimes: number[] = createArray(
    child.props.children.props.children.length
  )

  const TableLoader = (): JSX.Element => {
    return (
      <>
        <thead>
          <tr>
            {colTimes.map((colItem) => (
              <th key={`colItem_${colItem}`}>
                <Skeleton />
              </th>
            ))}
          </tr>
        </thead>
        <tbody>
          {rowTimes.map((rowItem) => (
            <tr key={`rowItem_${rowItem}`}>
              {colTimes.map((colItem) => (
                <td key={`colItem_${colItem}`}>
                  <Skeleton />
                </td>
              ))}
            </tr>
          ))}
        </tbody>
      </>
    )
  }

  return (
    <TableWrapper>
      <table>{loading ? <TableLoader /> : children}</table>
    </TableWrapper>
  )
}

export { Table }
