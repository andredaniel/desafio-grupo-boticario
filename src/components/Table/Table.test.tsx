import { render } from '@testing-library/react'
import React from 'react'
import { Table } from './Table'

describe('<Table />', () => {
  it('should render the Table component', () => {
    const { container } = render(
      <Table>
        <thead>
          <tr>
            <th>Header</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>Body</td>
          </tr>
        </tbody>
      </Table>
    )
    expect(container).toMatchSnapshot()
  })

  it('should render the Table component in load state', () => {
    const { container } = render(
      <Table loading={true}>
        <thead>
          <tr>
            <th>Header</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>Body</td>
          </tr>
        </tbody>
      </Table>
    )
    expect(container).toMatchSnapshot()
  })
})
