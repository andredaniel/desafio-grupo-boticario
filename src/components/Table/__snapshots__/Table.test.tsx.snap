// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should render the Table component 1`] = `
<div>
  <div
    class="sc-bdfBwQ hVYMEO"
  >
    <table>
      <thead>
        <tr>
          <th>
            Header
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Body
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<Table /> should render the Table component in load state 1`] = `
<div>
  <div
    class="sc-bdfBwQ hVYMEO"
  >
    <table>
      <thead>
        <tr />
      </thead>
      <tbody>
        <tr />
        <tr />
        <tr />
        <tr />
        <tr />
        <tr />
        <tr />
        <tr />
      </tbody>
    </table>
  </div>
</div>
`;
